### PR TITLE
Add multiscale structural similarity

### LINKED ISSUE #5194: Feature request: MS-SSIM calculation support in skimage.metrics
## Description

Hi all, I am wondering if I can use skimage to produce MS-SSIM between two images. 

Sadly, I find skimage still cannot support that. (Please correct me if I am wrong, or if we can calculate MS-SSIM easily by ourselves without the help of APIs. )

My feature request here is that to add MS-SSIM calculation API in `skimage.metrics`.

### PR DESCRIPTION
**Pull request by a-parida12**
## Description

Duplicate PR of  #6470  Closes #5194
Adding `multiscale_structural_similarity` to `skimage.metrics`

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Review by rfezzani — COMMENTED** [review]
Thank you @a-parida12 for your contribution. You can find my comments below. My major concern is about the multichannel support: you may add the `multichannel_axis` to the required argument.

**Comment by a-parida12** [comment]
@rfezzani thank you much for the reviews, comments and links. I am working on it!

File: skimage/metrics/_structural_similarity.py [hunk 1]
```diff
@@ -252,3 +270,62 @@ def structural_similarity(im1, im2,
...
+    - kwargs can be used to pass arguments for the ssim metric at each scale see metrics.ssim for details
+    - Code adapted version of the pytorch implementation of the msssim from `pytorch-mssim` by @jorge-pessoa       
+
+    """
+
+    check_shape_equality(im1, im2)
+    min_size_img = 2 ** len(multiscale_weights)
+
+    # assuming 3rd dim is the channel
+    assert min(im1.shape[0], im1.shape[1]), f"min required image size is {min_size_img}, multiscale_weights of length {len(multiscale_weights)} is being used"
```
[the remark this answers is not in the document]
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py:309]: how do I handle this? Do I truncate the weights to a compatible number and raise the warning?

My reason for the assertion error was to let the user update the values. Or is it too much?

[the remark this answers is not in the document]
Comment (rfezzani) [reply skimage/metrics/_structural_similarity.py]: > how do I handle this? Do I truncate the weights to a compatible number and raise the warning?

That sounds a good approach :wink: 

> My reason for the assertion error was to let the user update the values. Or is it too much?

The error assertion can be annoying for example when a user calls the function in a loop over images with different sizes: the user will have to provide different weights depending on image size :confused:

File: skimage/metrics/_structural_similarity.py [hunk 2]
```diff
@@ -252,3 +270,62 @@ def structural_similarity(im1, im2,
             return mssim, S
         else:
             return mssim
+
+
+@utils.deprecate_multichannel_kwarg()
+def multiscale_structural_similarity(im1, im2,
+                          *,
+                          win_size=11,
+                          multiscale_weights=[0.0448, 0.2856, 0.3001, 0.2363, 0.1333], **kwargs):
```
[the remark this answers is not in the document]
Comment (rfezzani) [reply skimage/metrics/_structural_similarity.py]: @a-parida12 can you please give some more details about the default weights values :pray:

[the remark this answers is not in the document]
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py]: https://ece.uwaterloo.ca/~z70wang/publications/msssim.pdf
Page 3 just before the results section

[the remark this answers is not in the document]
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py]: I will put a dev comment as well

File: skimage/metrics/_structural_similarity.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py]: Thanks! for the confirmation. I have done some changes now.

File: skimage/metrics/_structural_similarity.py [hunk 3]
```diff
@@ -252,3 +270,72 @@ def structural_similarity(im1, im2,
             return mssim, S
         else:
             return mssim
+
+
+def multiscale_structural_similarity(im1, im2,
+                          *,
+                          win_size=11,
+                          multiscale_weights=(0.0448, 0.2856, 0.3001, 0.2363, 0.1333),
+                          # \alpha, \beta values from https://ece.uwaterloo.ca/~z70wang/publications/msssim.pdf
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:279]: A line in the *Notes* docstring section is may be better to make this information available in the function documentation. It also allows to add the paper reference :wink:

File: skimage/metrics/_structural_similarity.py [hunk 4]
```diff
@@ -252,3 +270,72 @@ def structural_similarity(im1, im2,
...
+
+    """
+    check_shape_equality(im1, im2)
+
+    mssim = []
+    mcs = []
+    
+    for weight in multiscale_weights:
+        # calculate ssim at current scale
+        if min(im1.shape[0], im1.shape[1]) > 1:
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:319]: shouldn't `win_size` be the smallest allowed image size?

File: skimage/metrics/_structural_similarity.py [hunk 5]
```diff
@@ -252,3 +270,72 @@ def structural_similarity(im1, im2,
...
         else:
             return mssim
+
+
+def multiscale_structural_similarity(im1, im2,
+                          *,
+                          win_size=11,
+                          multiscale_weights=(0.0448, 0.2856, 0.3001, 0.2363, 0.1333),
+                          # \alpha, \beta values from https://ece.uwaterloo.ca/~z70wang/publications/msssim.pdf
+                          channel_axis = 2,
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:280]:
```suggestion
                          channel_axis=None,
```
by default, we usually assume that the image is grey level :wink:

File: skimage/metrics/_structural_similarity.py [hunk 6]
```diff
@@ -252,3 +270,72 @@ def structural_similarity(im1, im2,
...
+        # calculate ssim at current scale
+        if min(im1.shape[0], im1.shape[1]) > 1:
+            sim, cs = structural_similarity(im1, im2, win_size=win_size, channel_axis=channel_axis, full=True, **kwargs)
+            
+            # multiply the weights for current scale
+            mssim.append(sim ** weight)
+            mcs.append(cs.mean() ** weight)
+
+            # rescale the images for scale i+1
+            im1 = zoom(im1, zoom=0.5)
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:327]: What about exposing the scale factor and default it to 0.5?

File: skimage/metrics/_structural_similarity.py [hunk 7]
```diff
@@ -252,3 +270,72 @@ def structural_similarity(im1, im2,
...
+            im2 = zoom(im2, zoom=0.5)
+        else:
+            warn(
+                f"Running truncated mssim. To run full ms-ssim expected minimum img spatial dim {2 ** len(multiscale_weights)}",
+                stacklevel=2,
+            )
```
Comment (rfezzani) [inline skimage/metrics/_structural_similarity.py:333]:
```suggestion
            warn(
                f"Running truncated mssim. To run full ms-ssim expected 
                f"minimum image size {2 ** len(multiscale_weights)}",
                stacklevel=2,
            )
            break
```
The loop can be stopped in that case :wink:

Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py:333]: I think that is nice!

File: skimage/metrics/_structural_similarity.py (hunk 4 above)
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py]: Oops! absolutely you are right!

File: skimage/metrics/_structural_similarity.py (hunk 3 above)
Comment (a-parida12) [reply skimage/metrics/_structural_similarity.py]: makes sense!

**Pull request closed without merging**